java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestItemUtils.java

```diff
@@ -22,47 +22,53 @@
+import org.eclipse.jdt.core.SourceRange;
 import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.hover.JavaElementLabels;
 import org.eclipse.lsp4j.Range;
 
 @SuppressWarnings("restriction")
 public class TestItemUtils {
 
     public static final String DEFAULT_PACKAGE_NAME = "<Default Package>";
 
-    public static TestItem constructTestItem(IJavaElement element, TestLevel level) throws JavaModelException {
-        return constructTestItem(element, level, null);
-    }
-
-    public static TestItem constructTestItem(IJavaElement element, TestLevel level, TestKind kind)
+    public static JavaTestItem constructJavaTestItem(IJavaElement element, TestLevel level, TestKind kind)
             throws JavaModelException {
         final String displayName;
         final String fullName;
         if (element instanceof IPackageFragment && ((IPackageFragment) element).isDefaultPackage()) {
             displayName = DEFAULT_PACKAGE_NAME;
-            fullName = DEFAULT_PACKAGE_NAME;
+            fullName = "";
         } else {
-            displayName = element.getElementName();
-            fullName = parseTestItemFullName(element, level);
+            displayName = JavaElementLabels.getElementLabel(element, JavaElementLabels.ALL_DEFAULT);
+            fullName = parseFullName(element, level);
         }
         final String uri = JDTUtils.getFileURI(element.getResource());
-        final Range range = parseTestItemRange(element);
+        Range range = null;
+        if (level == TestLevel.CLASS || level == TestLevel.METHOD) {
+            range = parseTestItemRange(element);
+        }
+        
         final String projectName = element.getJavaProject().getProject().getName();
 
-        return new TestItem(displayName, fullName, uri, projectName, range, level, kind);
+        final JavaTestItem result = new JavaTestItem(displayName, fullName, projectName, uri, range, level, kind);
+        result.setJdtHandler(element.getHandleIdentifier());
+        return result;
     }
 
     public static Range parseTestItemRange(IJavaElement element) throws JavaModelException {
         if (element instanceof ISourceReference) {
-            // getSourceRange() is not used here because we want the Code Lens appear above the
-            // method name, not above the annotation.
-            final ISourceRange range = ((ISourceReference) element).getNameRange();
-            return JDTUtils.toRange(element.getOpenable(), range.getOffset(), range.getLength());
+            final ISourceRange sourceRange = ((ISourceReference) element).getSourceRange();
+            final ISourceRange nameRange = ((ISourceReference) element).getNameRange();
+            if (SourceRange.isAvailable(sourceRange) && SourceRange.isAvailable(nameRange)) {
+                return JDTUtils.toRange(element.getOpenable(), nameRange.getOffset(),
+                        sourceRange.getLength() - nameRange.getOffset() + sourceRange.getOffset());
```

Comment:
`sourceRange.getLength() - nameRange.getOffset() + sourceRange.getOffset()` equals to `nameRange.getLength()`?

can you briefly explain?

Reply:
sourceRange contains JavaDocs, so `(nameRange.getOffset() - sourceRange.getOffset())` equals to the length of JavaDoc, I'll add a comment for it.